apps/api-extractor/src/extractor/Extractor.ts

```diff
@@ -164,11 +164,50 @@ export class Extractor {
    * @param options - IExtractor options.
    */
   public static processProjectFromConfigFile(jsonConfigFile: string, options?: IExtractorOptions): void {
-    const configObject: IExtractorConfig = JsonFile.loadAndValidate(jsonConfigFile, Extractor.jsonSchema);
+    const configObject: IExtractorConfig = this.loadConfigObject(jsonConfigFile);
     const extractor: Extractor = new Extractor(configObject, options);
     extractor.processProject();
   }
 
+  /**
+   * Loads the api extractor config file in Extractor Config object.
+   * @param jsonConfigFile - Path to api extractor json config file.
+   */
+  public static loadConfigObject(jsonConfigFile: string): IExtractorConfig {
+    // Set to keep track of config files which have been processed.
+    const pathSet: Set<string> = new Set<string>();
+    // Get absolute path of config file.
+    let currentConfigFilePath: string = path.resolve(jsonConfigFile);
+    pathSet.add(currentConfigFilePath);
+
+    let extractorConfig: IExtractorConfig = JsonFile.load(jsonConfigFile);
+
+    while (extractorConfig.extends) {
+      // Populate the api extractor config path defined in extends relative to current config path.
+      currentConfigFilePath = path.resolve(path.dirname(currentConfigFilePath), extractorConfig.extends);
+
+      // Check if this file was already processed.
+      if (pathSet.has(currentConfigFilePath)) {
+        throw new Error('The api extractor config files contains a cycle. '
+        + currentConfigFilePath + ' is being processed again. '
+        + 'Please check the extends values in config files.');
+      }
+      pathSet.add(currentConfigFilePath);
+
+      // Remove extends property from config for current config.
+      delete extractorConfig.extends;
+
+      // Load the extractor config defined in extends property.
+      const baseConfig: IExtractorConfig = JsonFile.load(currentConfigFilePath);
+      lodash.merge(baseConfig, extractorConfig);
+      extractorConfig = baseConfig;
+    }
+
+    // Validate if the extractor config generated adheres to schema.
+    Extractor.jsonSchema.validateObject(extractorConfig, jsonConfigFile);
+    return extractorConfig;
```

Comment:
We should do this on every iteration so we can tell the user where the problem arises.

Reply:
I don't think the object will actually conform to the schema until all the `extends` objects have been merged.

---
In reply to: [236478087](https://github.com/Microsoft/web-build-tools/pull/957#discussion_r236478087) [](ancestors = 236478087)

Reply:
In case we want to do validation in each iteration we might have to keep separate schema file where all the properties are optional. In that case we can catch error if any invalid property is specified and display that file. In the end we can validate with current schema file. Not sure if that level of reporting is required.

Reply:
I'm not sure I'd want to maintain two different versions of the schema.  But maybe the schema validator has a way to disable required properties when validating partial input files.

Let's consider this as a separate PR, rather than blocking this PR.

---
In reply to: [236567986](https://github.com/Microsoft/web-build-tools/pull/957#discussion_r236567986) [](ancestors = 236567986)